Bug-62055 - Fix XSSFImportFromXML table resize

kapexx opened the pull request:
Fixes [Bug-62055](https://bz.apache.org/bugzilla/show_bug.cgi?id=62055) and one other issue with `XSSFImportFromXML`.

The main change is that importing `XSSFImportFromXML.importFromXML` now resizes mapped tables when importing XML data. Each table will have the a row count of exactly the amount of imported elements, i.e. it shrinks or grows accordingly. When growing, cells below the table are overwritten.

The other issue this fixes is that table column id's were wrongly used to get the column position. This resulted in data written to wrong columns when the id did not match the positions.

Additional checks have been added to the tests for both issues. Changes to `CustomXMLMappings-complex-type.xlsx` include: 
- Added an additional (unmapped) column to verify that data is written to correct columns when the column id doesn't match the columns position
- Added some text in a cell below a table that will grow, to check if the cell is correctly overwritten when the table grows
- Added some text in a cell inside a table that will shrink, to check if the cell is correctly cleared when the the table shrinks

Commit:
Bug 62055: Resize tables when importing XML map

asfgit commented:
Can one of the admins verify this patch?

pjfanning commented on src/ooxml/java/org/apache/poi/xssf/extractor/XSSFImportFromXML.java:
```diff
@@ -87,25 +91,25 @@ public XSSFImportFromXML(XSSFMap map) {
      */
     public void importFromXML(String xmlInputString) throws SAXException, XPathExpressionException, IOException {
-
+    
```
Could you remove the code formatting changes?

pjfanning commented on src/ooxml/testcases/org/apache/poi/xssf/extractor/TestXSSFImportFromXML.java:
```diff
@@ -98,9 +100,10 @@ public void testMultiTable() throws IOException, XPathExpressionException, SAXEx
+			
+			int cellOffset = 10; // cell C10
+			for (int i = 0; i < 10000; i++) {
+				testXML.append("<ns1:Schema ID=\"c").append(i + cellOffset).append("\" SchemaRef=\"d\" />");
```
could you use spaces for indents?

pjfanning commented on src/ooxml/java/org/apache/poi/xssf/extractor/XSSFImportFromXML.java:
```diff
@@ -160,6 +172,57 @@ public void importFromXML(String xmlInputString) throws SAXException, XPathExpre
+        }
+        
+        // update table area
+        table.getCTTable().setRef(newTableAreaReference.formatAsString());
```
We regard accessing the CT* generated classes as something we want to discourage.
Could you change the the XSSFTable class to expose a setRef method?

kapexx replied:
I just noticed `setRef` is already exposed as `XSSFTable.setCellReferences`, so I'll use that.

I also had to use CTTable and CTTableColumn near line 143 though, to get the correct column index. Now I saw that I could have used `XSSFTable.findColumnIndex(String columnHeader)` instead of calculating the column index myself, but it seems I would still have to use `CTTableColumn` to get the column header name. 

I didn't want to make any API changes before, but I guess in this case it would be useful if `XSSFXmlColumnPr` exposes the column header name and maybe also provides its column index. That would avoid using any CT* classes. What do you think?

pjfanning replied:
If you could add any APIs that you need, that would be great. We want to make the generated CT* classes only internal implementation details (allowing us to move away from xmlbeans code generation eventually).

Alain-Bearez replied:
There is an ongoing effort to replace calls to the underlying `CT*` classes, so that these calls be limited to the class directly wrapping the XML object. 

Every attempt in that direction is welcome.

Alain-Bearez commented on src/ooxml/java/org/apache/poi/xssf/extractor/XSSFImportFromXML.java:
```diff
@@ -160,6 +172,57 @@ public void importFromXML(String xmlInputString) throws SAXException, XPathExpre
+        
+        table.updateReferences();
+        int rowCount = table.getRowCount();
+        if(rowCount == newRowCount) {
```
The project uses space between keywords (if, for, try, while) and opening parenthesis.

Alain-Bearez commented on src/ooxml/testcases/org/apache/poi/xssf/extractor/TestXSSFImportFromXML.java:
```diff
@@ -85,7 +87,7 @@ public void testImportFromXML() throws IOException, XPathExpressionException, SA
 	}
 	
-	@Test(timeout=60000)
+	@Test(timeout=60000_00)
```
Why multiply by 100?

kapexx replied:
Sorry, that was a hack to avoid that the test is terminated to early while debugging. I forgot to change it back. I know that JUnit Timeout rules can be deactivated while in debug mode, but I didn't find any way to do the same for timeouts defined in the annotation.

Commit:
Fix formatting and remove debug code

- Convert tabs to spaces in TestXSSFImportFromXML
- Remove spaces from otherwise empty lines
- Add space after if/for keywords
- Remove accidentally left code for debugging

Commit:
Remove usage of CTTable and CTTableColumn, expose column name in XSSFXmlColumnPr

pjfanning commented on src/ooxml/java/org/apache/poi/xssf/usermodel/helpers/XSSFXmlColumnPr.java:
```diff
@@ -32,56 +31,60 @@ Licensed to the Apache Software Foundation (ASF) under one or more
+    }
+
+    /**
+     * @return the column name
```
could you add `@since 4.0.0` to the javadoc for new API methods?

kapexx replied:
I reconsidered adding the method in `XSSFXmlColumnPr` because I think it actually belongs to `XSSFTableColumn` (which didn't exist yet, so I've created it - see next commit)

Commit:
Add XSSFTableColumn as wrapper for CTTableColumn

Rollback changes to XSSFXmlColumnPr, move getName() to XSSFTableColumn

kapexx commented:
I have created the class `XSSFTableColumn` as wrapper for `CTTableColumn`, which was missing. It exposes the column id, name and `XSSFXmlColumnPr` (if the column has an XML mapping). In this basic version `XSSFTableColumn` only makes it a bit easier to handle non-XML-mapped columns but the class could later be improved to support other column features like column formulas, header/data cell styles, etc.

Commit:
Deprecate methods in XSSFXmlColumnPr that have been moved to XSSFTableColumn

- Remove usage of CTTableColumn from XSSFXmlColumnPr and XSSFExportToXml

pjfanning reviewed (approved):
LGTM

pjfanning commented:
@Alain-Bearez are you happy for me to merge the changes?

Alain-Bearez commented on src/ooxml/java/org/apache/poi/xssf/usermodel/XSSFTableColumn.java:
```diff
@@ -0,0 +1,109 @@
+     *            the table column xmlbean to wrap
+     * @since 4.0.0
+     */
+    public XSSFTableColumn(XSSFTable table, CTTableColumn ctTableColumn) {
```
I would insist in having all methods that use a `CT*` parameter to be annotated as `@Internal` when they are not private.

kapexx replied:
I'll check the code and add those annotations.

This particular constructor doesn't seem to be useful for the user anyway. I think I could make this constructor's visibility be `protected` instead of `public` and then cleanup `XSSFTable.addColumn` and make it part of the public API. `XSSFTable.addColumn` then is the correct and only way to create new table columns in the API.

Alain-Bearez replied:
Definitively the best option!

Alain-Bearez reviewed (commented):
My only finding after re-reviewing the changes is related to the (lack of) use of `@Internal` annotation for methods the use internal `CT*` parameters.

Maybe this should be more coherently handled as a distinct global task, all over the `ooxml` code base.

Commit:
Add XSSFTable column creation a table area methods

- Move table resize method from XSSFImportFromXML to XSSFTable

XSSFTable:
- Make createColumn part of the public API
- Add removeColumn methods
- Add getColumnCount
- Add getArea/setArea to handle the table's area
- Add getDataRowCount/setDataRowCount to handle row count without including header/totals rows

Deprecate some methods related to table size and columns:
- Deprecate XSSFTable.getXmlColumnPrs which is replaced by XSSFTableColumn.getXmlColumnPr
- Deprecate XSSFTable.getNumberOfMappedColumns which is replaced by getColumnCount
- Deprecate XSSFTable.setCellReferences which can easily lead to corrupted tables. Internally replaced by setCellRef, in the public API replaced by setArea

kapep committed:
Merge branch 'trunk' into Bug62055_XSSFImportFromXML-table-resize

kapexx commented:
I moved the table resizing from the XML import into the XSSFTable. I've gone a step further and implemented some more methods related to the table area, rows and columns. I deprecated and replaced a few methods or made them internal, because those methods easily lead to corrupted tables e.g. if the user forgets set the table's area or to add columns manually after setting the tables area. The replacement methods do some additional verifications and for example automatically add or remove columns to avoid corrupted tables.
